modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/RebalanceRaftGroupEventsListener.java

```diff
@@ -103,15 +136,75 @@ public void onNewPeersConfigurationApplied(List<PeerId> peers) {
 
         try {
             doOnNewPeersConfigurationApplied(peers);
+
+            rebalanceAttempts.set(0);
         } finally {
             busyLock.leaveBusy();
         }
     }
 
     /** {@inheritDoc} */
     @Override
-    public void onReconfigurationError(Status status) {
-        // TODO: IGNITE-14873 implement this method
+    public void onReconfigurationError(Status status, Supplier<Void> rebalanceRunner) {
+        if (!busyLock.enterBusy()) {
+            return;
+        }
+
+        try {
+            if (status == null) {
+                // leader stepped down, so we are expecting RebalanceRaftGroupEventsListener.onLeaderElected to be called on a new leader.
+                LOG.info("Leader stepped down during the current rebalance for the partId = {}.");
+
+                return;
+            }
+
+            assert status.getRaftError() == RaftError.ECATCHUP : "According to the JRaft protocol, RaftError.ECATCHUP is expected.";
+
+            LOG.warn("Error occurred during the current rebalance for partId = {}.", partId);
+
+            if (rebalanceAttempts.incrementAndGet() < REBALANCE_RETRY_THRESHOLD) {
+                rebalanceScheduler.schedule(() -> {
+                    LOG.info("Started {} attempt to retry the current rebalance for the partId = {}.", rebalanceAttempts.get(), partId);
+
+                    rebalanceRunner.get();
+                }, REBALANCE_RETRY_DELAY_MS, TimeUnit.MILLISECONDS);
+            } else {
+                LOG.info("Failed to perform the current rebalance for the partId = {}, canceling.", partId);
+
+                cancelRebalance();
+
+                rebalanceAttempts.set(0);
+            }
+        } finally {
+            busyLock.leaveBusy();
+        }
+
+    }
+
+    /**
+     * This is a draft version of canceling a rebalance, this approach contains some serious flaws and must be revised.
+     */
+    private void cancelRebalance() {
+        Entry plannedEntry = metaStorageMgr.get(plannedPartAssignmentsKey(partId)).join();
+
+        if (plannedEntry.value() != null) {
```

Comment:
We also should add an atomic logic of checking whether pending.revision is the same as the one we try to cancel, otherwise we might cancel new one.

Imaging following situation:
PENDING=[X]
Node A tries to rebalance data to X and fails all n retries -> it decides to cancel such X, thread sleep.
New Leander B elected that detects pending X and succeeds with it. On success it moves planned Y to pending.
Node A rebalance thread wakes up and cancels pending Y.

Reply:
We discussed this approach with @kgusakov and we came to the conclusion, that is more preferable to throw some exception, when we haven't succeeded in the evaluation of the rebalance, instead of canceling it by removing `pending key`. We agreed that the current solution with removing pending key is not completed and may cause inconsistency in the current rebalance protocol. 

Also there is a case, when we can lost planned rebalance: 
1) Current leader retries failed rebalance
2) Current leader stops being leader for some reasons and sleeps
3) New leader performs rebalance and calls `RebalanceRaftGroupEventsListener#onNewPeersConfigurationApplied`
4) At this moment old leader wakes up and cancels the current rebalance, so it removes pending and writes to it planned key.
5) At this moment we receive `RebalanceRaftGroupEventsListener#onNewPeersConfigurationApplied` from the new leader, see that planned is empty, so we just delete pending key, but this is not correct to delete this key as far as the rebalance that is associated to the removed key hasn't been performed yet.

To sum up, I believe that on the current state of the rebalance feature, where we don't have proper failure handler, it is more convenient to throw an exception, so we can immediately react to  exceptional scenario on the TC, when we couldn't perform rebalance even after several retires.

@sanpwc WDYT?

Reply:
We have discussed the solution in person and agreed to have infinite retry mechanism at this point of development, and to create separate ticket for rebalance canceling